### PR TITLE
test Swift 2.3 & 3.0 in .jenkins.yml

### PR DESCRIPTION
**Pull request by jpsim**
Not ready for review yet.

### CODE REVIEW

**Comment by jpsim** [comment]
I've added code that will detect common Xcode/Simulator errors that we encounter and rerun the job, since those issues are rare enough that they typically don't happen twice in a row. You can see this in action here: https://ci.realm.io/job/objc_pr/4048/configuration=Debug,swift_version=3.0,target=ios-dynamic/console

```
▸ Signing build/DerivedData/Realm/Build/Products/Debug-iphonesimulator/TestHost.app/PlugIns/Tests.xctest
2016-09-12 09:20:26.236 xcodebuild[86849:647471] Connection peer refused channel request for "dtxproxy:XCTestManager_IDEInterface:XCTestManager_DaemonConnectionInterface"; channel canceled <DTXChannel: 0x7ff7236fa910>
2016-09-12 09:20:26.335 xcodebuild[86849:647466] Error Domain=IDETestOperationsObserverErrorDomain Code=3 "The operation couldn’t be completed. (DTXProxyChannel error 1.) If you believe this error represents a bug, please attach the log file at /Users/realm/workspace/objc_pr/configuration/Debug/swift_version/3.0/target/ios-dynamic/build/DerivedData/Realm/Logs/Test/7FE2575E-18E0-4FC8-887E-DFF02452E3D4/Session-Tests-2016-09-12_092002-6S6d0g.log" UserInfo={NSLocalizedDescription=The operation couldn’t be completed. (DTXProxyChannel error 1.) If you believe this error represents a bug, please attach the log file at /Users/realm/workspace/objc_pr/configuration/Debug/swift_version/3.0/target/ios-dynamic/build/DerivedData/Realm/Logs/Test/7FE2575E-18E0-4FC8-887E-DFF02452E3D4/Session-Tests-2016-09-12_092002-6S6d0g.log}
Known Xcode error detected. Running job again.
```

**Comment by jpsim** [comment]
This is ready for review @bdash @tgoyne @austinzheng.

I'd like to re-enable the skipped Xcode 8 `build.sh` jobs in a future PR.

**Comment by austinzheng** [comment]
👍

File: scripts/reset-simulators.sh [hunk 1]
```diff
@@ -1,59 +1,43 @@
...
-    pkill Simulator 2>/dev/null || true
+    pkill Simulator 2>/dev/null
     # CoreSimulatorService doesn't exit when sent SIGTERM
-    pkill -9 Simulator 2>/dev/null || true
-  done
+    pkill -9 Simulator 2>/dev/null
+done
+
+# Install jq if needed
+if ! which jq >/dev/null; then
```
Comment (tgoyne) [inline scripts/reset-simulators.sh]: Shouldn't need to leave this here.

File: scripts/reset-simulators.sh [hunk 2]
```diff
@@ -1,59 +1,43 @@
 #!/bin/bash
 
-set -o pipefail
```
Comment (tgoyne) [inline scripts/reset-simulators.sh]: Just stumbling on through when something fails in an unexpected way seems undesirable.

File: scripts/reset-simulators.sh (hunk 1 above)
Comment (jpsim) [inline scripts/reset-simulators.sh]: removed

File: scripts/reset-simulators.sh (hunk 2 above)
Comment (jpsim) [inline scripts/reset-simulators.sh]: re-added

File: scripts/reset-simulators.sh [hunk 3]
```diff
@@ -1,59 +1,38 @@
...
-            previous_device=""
-            continue
-        fi
-
-        regex='^(.*) [(]([0-9A-F]{8}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{12})[)]'
-        if [[ $LINE =~ $regex ]]; then
-            device="${BASH_REMATCH[1]}"
-            guid="${BASH_REMATCH[2]}"
+# Shut down booted simulators
+xcrun simctl list devices -j | jq -c -r '.devices | flatten | map(select(.availability == "(available)")) | map(select(.state == "Booted")) | map(.udid) | .[]' | while read udid; do
```
Comment (bdash) [inline scripts/reset-simulators.sh]: Does this do something different than the old code, or is it just a more concise way of achieving the same thing? Relying on a third-party tool being installed for this script to work makes me a little sad.

Comment (jpsim) [inline scripts/reset-simulators.sh]: It does something different than the old code. The textual output format used by `simctl` changed between Xcode 7 and Xcode 8, and accommodating for those differences is cumbersome, and fragile.

**Comment by tgoyne** [comment]
The one test failure (`-[RealmSwift_Tests.RealmTests testRefresh]` in Swift 3.0 on iOS Release) is not one that I've seen before. The actual failure is that `fulfill` is being called twice, and the failure after that is just due to the test not completing as a result.

**Comment by jpsim** [comment]
Any ideas on how to resolve the `testRefresh` failure @tgoyne?

**Comment by jpsim** [comment]
I've added a number of Swift 2.3 & 3.0 installation and usage examples.

**Comment by jpsim** [comment]
The `testRefresh` failure seems fairly rare, I don't think it's necessary to address here.

@bdash / @tgoyne any further comments?

**Comment by jpsim** [comment]
Bump @tgoyne @bdash. Would like to get this in prior to landing Swift 3 API changes.

**Comment by tgoyne** [comment]
👍

**Pull request merged**